backend/siarnaq/bracket/challonge.py

```diff
@@ -0,0 +1,264 @@
+    tournament_challonge_id = (
+        tournament.external_id_private if is_private else tournament.external_id_public
+    )
+
+    url = f"{URL_BASE}tournaments/{tournament_challonge_id}.json"
+
+    r = requests.get(url, headers=_headers)
+    r.raise_for_status()
+    return r.json()
+
+
+def get_round_indexes(tournament: Tournament, *, is_private: bool):
+    tournament_data = get_tournament_data(tournament, is_private=is_private)
+
+    round_indexes = set()
+    for item in tournament_data["included"]:
+        match item:
+            case {"type": "match", "attributes": {"round": round_index}}:
+                round_indexes.add(round_index)
+
+    return round_indexes
+
+
+def get_match_and_participant_objects_for_round(
+    round: TournamentRound, *, is_private: bool
+):
+    tournament_data = get_tournament_data(round.tournament, is_private=is_private)
+    # Derive match dicts/JSON objects (that Challonge gives us) of this round
+    challonge_matches = []
+    # Also derive participant dicts/JSON objects that Challonge gives us,
+    # and map them with IDs for easy lookup
+    challonge_participants = dict()
+
+    for item in tournament_data["included"]:
+        match item:
+            case {
+                "type": "match",
+                "attributes": {"round": round.external_id, "state": state},
+            }:
+                # Only enqueue the round if all matches are "open".
+                # NOTE: it would be good to have a "force re-enqueue round",
+                # which re-enqueues matches even if matches or round
+                # already in progress.
+                # This would change the following check --
+                # matches could be open _or done_.
+                # !!! This is also _really hard_ right now
+                # cuz it involves match deletion which is really hard.
+                # Track in #594
+                if state != "open":
+                    raise RuntimeError(
+                        "The bracket service's round does not only\
+                            have matches that are ready."
+                    )
+                challonge_matches.append(item)
+
+            case {"type": "participant", "id": id}:
+                challonge_participants[id] = item
+
+    match_objects = []
+    match_participant_objects = []
+
+    for challonge_match in challonge_matches:
+        match_object = apps.get_model("compete", "Match")(
+            episode=round.tournament.episode,
+            tournament_round=round,
+            alternate_order=True,
+            is_ranked=False,
+            challonge_id=challonge_match["id"],
+        )
+        match_objects.append(match_object)
+
+        # Note that Challonge 1-indexes its player indexes
+        # while Saturn 0-indexes.
+        for (challonge_player_index, saturn_player_index) in (
+            ("player1", 0),
+            ("player2", 1),
+        ):
```

Comment:
Not Saturn. This is purely to do with the design of our database layout. Also please rename `saturn_player_index`.

```suggestion
        # while our data model 0-indexes.
        for (challonge_player_index, saturn_player_index) in enumerate(
            ["player1", "player2"]
        ):
```